File: csharp/Tdcv2/Format/Interpolate.cs
using System.Collections.Concurrent;
using System.Text;
using System.Text.RegularExpressions;

namespace Tdcv2.Format;

/// <summary>
/// <c>${{Name}}</c> and <c>${{Name|upper|mask:xxx}}</c> inside a <c>&lt;data&gt;</c>.
/// </summary>
/// <remarks>
/// <para>
/// The marker itself is configurable through <c>&lt;env inject="..."&gt;</c>: the <c>%</c> in it
/// stands for the name, and everything around it is the delimiter. A config generating shell
/// scripts can set <c>inject="&lt;&lt;%&gt;&gt;"</c> and stop fighting with dollar signs.
/// </para>
/// <para>
/// A name that matches no sequence is left exactly as it was written, marker and all. Replacing it
/// with an empty string would hide a typo inside data that still looks well-formed; leaving
/// <c>${{Gendre}}</c> in the output makes it obvious on the first row.
/// </para>
/// </remarks>
public static class Interpolate
{
    /// <summary>Reads the <c>inject</c> attribute; the greedy group picks the rightmost usable <c>%</c>.</summary>
    private static readonly Regex InjectShape = new("(.+)%(.+)", RegexOptions.Compiled);

    // Null is a value here, not an absence: the cache has to remember "this inject has no slot" or
    // it would try to recompile that pattern on every single line.
    private static readonly ConcurrentDictionary<string, Regex?> PatternCache = new();

    /// <summary>What a name resolves to on the row being rendered.</summary>
    public interface ILookup
    {
        bool Has(string name);

        string Value(string name);
    }

    public static string Apply(string text, string? inject, ILookup lookup)
    {
        Regex? pattern = PatternFor(inject);
        if (pattern is null)
        {
            // An inject with no `%` names nothing, so there is nothing to substitute.
            return text;
        }

        var result = new StringBuilder();
        int last = 0;
        foreach (Match m in pattern.Matches(text))
        {
            result.Append(text, last, m.Index - last);
            Reference reference = ParseReference(m.Groups[1].Value);
            if (!lookup.Has(reference.Name))
            {
                result.Append(m.Value);
            }
            else
            {
                string value = lookup.Value(reference.Name);
                foreach (Filter f in reference.Filters)
                {
                    value = Transforms.ApplyFilter(f.Kind, f.Arg, value);
                }

                result.Append(value);
            }

            last = m.Index + m.Length;
        }

        result.Append(text[last..]);
        return result.ToString();
    }

    private sealed record Filter(string Kind, string? Arg);

    private sealed record Reference(string Name, IReadOnlyList<Filter> Filters);

    /// <summary>
    /// <c>NAME ( "|" filter )*</c>, where a filter is a bare word or <c>word:arg</c>.
    /// </summary>
    /// <remarks>
    /// The argument runs to the next <c>|</c>, which is why a mask pattern may contain anything but
    /// a pipe.
    /// </remarks>
    private static Reference ParseReference(string raw)
    {
        string[] parts = raw.Split('|');
        string name = parts[0].Trim();
        var filters = new List<Filter>();
        for (int i = 1; i < parts.Length; i++)
        {
            string piece = parts[i];
            int colon = piece.IndexOf(':');
            if (colon < 0)
            {
                string bare = piece.Trim();
                if (bare.Length > 0)
                {
                    filters.Add(new Filter(bare, null));
                }
            }
            else
            {
                string kind = piece[..colon].Trim();
                if (kind.Length > 0)
                {
                    filters.Add(new Filter(kind, piece[(colon + 1)..].Trim()));
                }
            }
        }

        return new Reference(name, filters);
    }

    /// <summary>Null when the inject has no <c>%</c> slot at all.</summary>
    private static Regex? PatternFor(string? inject)
    {
        string key = string.IsNullOrEmpty(inject) ? "${{%}}" : inject!;
        return PatternCache.GetOrAdd(key, k =>
        {
            Match shape = InjectShape.Match(k);
            if (!shape.Success || shape.Length != k.Length)
            {
                return null;
            }

            return new Regex(
                Regex.Escape(shape.Groups[1].Value) + "(.+?)" + Regex.Escape(shape.Groups[2].Value),
                RegexOptions.Compiled);
        });
    }
}

File: python/src/tdcv2/format/interpolate.py
"""``${{Name}}`` and ``${{Name|upper|mask:xxx}}`` inside a ``<data>``.

The marker itself is configurable through ``<env inject="...">``: the ``%`` in it stands for the
name and everything around it is the delimiter. A config generating shell scripts can set
``inject="<<%>>"`` and stop fighting with dollar signs.

A name that matches no sequence is left exactly as it was written, marker and all. Replacing it
with an empty string would hide a typo inside data that still looks well-formed; leaving
``${{Gendre}}`` in the output makes it obvious on the first row.
"""

from __future__ import annotations

import re
from collections.abc import Callable
from dataclasses import dataclass

from .transforms import apply_filter

DEFAULT_INJECT = "${{%}}"

# The greedy first group picks the RIGHTMOST usable `%`, which is what lets an exotic wrapper
# such as inject="%{%}%" hold a literal percent sign of its own.
_INJECT_SHAPE = re.compile(r"(.+)%(.+)")

_PATTERN_CACHE: dict[str, re.Pattern[str] | None] = {}
_TEMPLATE_CACHE: dict[str, list[str | _Reference]] = {}


@dataclass(frozen=True, slots=True)
class _Filter:
    kind: str
    arg: str | None


@dataclass(frozen=True, slots=True)
class _Reference:
    name: str
    filters: list[_Filter]
    original: str
    """The marker as written, so an unknown name can be put back untouched."""


def apply(template: str, inject: str, lookup: Callable[[str], str | None]) -> str:
    """``template`` with every known reference replaced by its value on this row.

    ``lookup`` answers ``None`` for a name it does not know — a typo, which stays visible — and a
    string for one it does, including the empty string for a row a parent filtered out.
    """
    segments = compile_template(template, inject)
    if len(segments) == 1 and isinstance(segments[0], str):
        return segments[0]

    out: list[str] = []
    for segment in segments:
        if isinstance(segment, str):
            out.append(segment)
            continue
        value = lookup(segment.name)
        if value is None:
            out.append(segment.original)
            continue
        for f in segment.filters:
            value = apply_filter(f.kind, f.arg, value)
        out.append(value)
    return "".join(out)


def compile_template(template: str, inject: str) -> list[str | _Reference]:
    """The template split into literals and references, once.

    The render loop asks for the same ``<data>`` text on every row, so compiling once turns a
    per-row regex substitution — a measured hot spot — into a plain join.
    """
    key = f"{inject}\x00{template}"
    cached = _TEMPLATE_CACHE.get(key)
    if cached is not None:
        return cached

    pattern = _pattern(inject)
    segments: list[str | _Reference] = []
    if pattern is None:
        # An inject with no `%` names nothing, so there is nothing to substitute.
        segments.append(template)
    else:
        last = 0
        for m in pattern.finditer(template):
            if m.start() > last:
                segments.append(template[last : m.start()])
            name, filters = parse_reference(m.group(1))
            segments.append(_Reference(name, filters, m.group()))
            last = m.end()
        if last < len(template):
            segments.append(template[last:])
        if not segments:
            segments.append("")
    _TEMPLATE_CACHE[key] = segments
    return segments


def parse_reference(raw: str) -> tuple[str, list[_Filter]]:
    """``NAME ( "|" filter )*``, where a filter is a bare word or ``word:arg``.

    The argument runs to the next ``|``, which is why a mask pattern may contain anything except a
    pipe.
    """
    parts = raw.split("|")
    name = parts[0].strip()
    filters: list[_Filter] = []
    for piece in parts[1:]:
        colon = piece.find(":")
        if colon < 0:
            kind = piece.strip()
            if kind:
                filters.append(_Filter(kind, None))
        else:
            kind = piece[:colon].strip()
            if kind:
                filters.append(_Filter(kind, piece[colon + 1 :].strip()))
    return name, filters


def _pattern(inject: str) -> re.Pattern[str] | None:
    """``None`` when the inject has no ``%`` slot at all."""
    key = inject if inject else DEFAULT_INJECT
    if key in _PATTERN_CACHE:
        return _PATTERN_CACHE[key]

    shape = _INJECT_SHAPE.fullmatch(key)
    compiled = (
        None
        if shape is None
        else re.compile(f"{re.escape(shape.group(1))}(.+?){re.escape(shape.group(2))}")
    )
    _PATTERN_CACHE[key] = compiled
    return compiled
